masq/src/terminal_interface.rs

```diff
@@ -0,0 +1,439 @@
+    Ok(MemoryTerminal::new())
+}
+
+impl Clone for TerminalWrapper {
+    fn clone(&self) -> Self {
+        Self {
+            interface: Arc::clone(&self.interface),
+        }
+    }
+}
+
+fn interface_configurator<F, U, E: ?Sized, D>(
+    interface_raw: Box<F>,
+    terminal_type: Box<E>,
+) -> Result<TerminalReal, String>
+where
+    F: FnOnce(&'static str, U) -> std::io::Result<D>,
+    E: FnOnce() -> std::io::Result<U>,
+    U: linefeed::Terminal + 'static,
+    D: InterfaceRaw + Send + Sync + 'static,
+{
+    let terminal: U = match terminal_type() {
+        Ok(term) => term,
+        Err(e) => return Err(format!("Local terminal: {}", e)),
+    };
+    let mut interface: Box<dyn InterfaceRaw + Send + Sync + 'static> =
+        match interface_raw("masq", terminal) {
+            Ok(interface) => Box::new(interface),
+            Err(e) => return Err(format!("Preparing terminal interface: {}", e)),
+        };
+
+    if let Err(e) = set_all_settable_or_give_an_error(&mut *interface) {
+        return Err(e);
+    };
+
+    Ok(TerminalReal::new(interface))
+}
+
+fn set_all_settable_or_give_an_error<U>(interface: &mut U) -> Result<(), String>
+where
+    U: InterfaceRaw + Send + Sync + 'static + ?Sized,
+{
+    if let Err(e) = interface.set_prompt(MASQ_PROMPT) {
+        return Err(format!("Setting prompt: {}", e));
+    }
+
+    //here we can add another parameter to be configured,
+    //such as "completer" (see linefeed library)
+
+    Ok(())
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+pub trait MasqTerminal {
+    fn provide_lock(&self) -> Box<dyn WriterLock + '_> {
+        intentionally_blank!()
+    }
+    fn read_line(&self) -> TerminalEvent {
+        intentionally_blank!()
+    }
+
+    #[cfg(test)]
+    fn test_interface(&self) -> MemoryTerminal {
+        intentionally_blank!()
+    }
+    #[cfg(test)]
+    fn tell_me_who_you_are(&self) -> String {
+        intentionally_blank!()
+    }
+}
+//declaration of TerminalReal is in line_reader.rs
+
+#[derive(Default)]
+pub struct TerminalInactive {}
+
+impl MasqTerminal for TerminalInactive {
+    fn provide_lock(&self) -> Box<dyn WriterLock + '_> {
+        Box::new(WriterInactive {})
+    }
+    #[cfg(test)]
+    fn tell_me_who_you_are(&self) -> String {
+        "TerminalIdle".to_string()
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+pub trait WriterLock {
+    //I failed in attempts to use 'Any' and dynamic casting from Box<dyn WriterGeneric>
+    //because: Writer doesn't implement Clone and many if not all methods of Any require
+    //'static, that is, it must be an owned object and I cannot get anything else but a referenced
+    //Writer there.
+    //For delivering at least some tests I decided to use this, kind of a hack.
+    #[cfg(test)]
+    fn tell_me_who_you_are(&self) -> String {
+        intentionally_blank!()
```

Comment:
I disagree, since one of three takers of this trait doesn't need it anywhere to use. I had to link in a comment explaining why those structures are so complicated. If I knew how to get rid of them I'd do that happily.